### ISSUE TITLE
Explore fixes to perf/memory regressions around cloning/mutating patterns in components::Bag

### ISSUE THREAD

**Issue by gregtatum** [unicode-org/icu4x#877]
In #832 I'm introducing the ability to mutate patterns generated by the component::Bag. Previously the patterns could be returned by reference, and now they are copied. This is going to hit us on all key perf / memory / size metrics that we care about.

There are some potential mitigations. @sffc suggests that we can include a mapper function in the result to map each PatternItem to a potentially new one. This would work fine for adjusting widths or hour cycles, but won't work for append items, such as adding time zones or weekdays.

**Comment by sffc** [comment]
There are two general approaches to problems like this.  One is to fully compute the pattern in the constructor, cache it, and use it when needed.  The other is to compute the pattern on the fly for every formatting operation.

With #832, you take the approach of pre-computing the pattern in the constructor and allocating memory to cache it.

It comes down to the speed/memory tradeoff.  I think it's worth exploring whether the improvements we get to memory usage in the second approach are worth whatever speed impacts we might take.

**Comment by gregtatum** [comment]
Does #1131 address this?

**Comment by zbraniecki** [comment]
No, it's a separate issue. #1131 is about avoiding allocations for Date+Time pattern, this is about applying timezone and hour_cycle overrides on patterns from DataProvider based on user preferences.

**Comment by zbraniecki** [comment]
My hunch is that Shane's idea is a worth exploring. We could instead of modifying patterns, have a switch in format method that uses user preferences to decide which symbol to print.
